File: 0139-word-break/0139-word-break.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        n = len(s)
        dp = [False for _ in range(n)]

        for i in range(len(s)):
            for word in wordDict:
                if i < len(word) - 1:
                    continue
                if i == len(word) - 1 or dp[i - len(word)] == True:
                    if s[i-len(word)+1:i+1] == word:
                        dp[i] = True 
                        break


        return dp[-1]
                

File: 0139-word-break/0139-word-break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        int n = s.length();
        boolean[] dp = new boolean[s.length()+1];
        dp[n] = true;

        for(int i=n-1; i>=0; i--){
            for(String word: wordDict){
                if(i + word.length() <= n && s.substring(i, i+word.length()).equals(word)){
                    dp[i] = dp[i+word.length()];
                }
                if(dp[i]) break;
            }
        }

        return dp[0];
    }
}